metaflow/plugins/parallel_decorator.py

```diff
@@ -25,6 +43,79 @@ def step_init(
         self, flow, graph, step_name, decorators, environment, flow_datastore, logger
     ):
         self.environment = environment
+        # We choose `setattr` instead of the default `current._update_env` for two reasons: first, to preserve
+        # the datatype of `current.parallel` as a namedtuple, and second, to ensure the contents of the namedtuple
+        # are accurately set based on when they are accessed in the Metaflow lifecycle. The values for `main_ip`,
+        # `num_nodes`, and `node_index` are determined by environment variables which maybe set in `task_pre_step`.
+        # Thus, we use a `property` to lazily evaluate these namedtuple values.
+        setattr(
+            current.__class__,
+            "parallel",
+            property(
+                fget=lambda _: Parallel(
+                    main_ip=os.environ.get("MF_PARALLEL_MAIN_IP", "127.0.0.1"),
+                    num_nodes=int(os.environ.get("MF_PARALLEL_NUM_NODES", "1")),
+                    node_index=int(os.environ.get("MF_PARALLEL_NODE_INDEX", "0")),
+                )
+            ),
+        )
+
+    def task_pre_step(
+        self,
+        step_name,
+        task_datastore,
+        metadata,
+        run_id,
+        task_id,
+        flow,
+        graph,
+        retry_count,
+        max_user_code_retries,
+        ubf_context,
+        inputs,
+    ):
+        from metaflow import current
+
+        # Set `is_parallel` to `True` in `current` just like we
+        # with `is_production` in the project decorator.
+        current._update_env(
+            {
+                "is_parallel": True,
+            }
+        )
+
+        self.input_paths = [obj.pathspec for obj in inputs]
+        if not hasattr(flow, "_parallel_ubf_iter"):
+            raise MetaflowException(
+                "Parallel decorator is only supported in unbounded foreach steps."
```

Comment:
curious - how is this error triggered? shouldn't the linter throw this error upstream?

Reply:
It shouldn't be; Added because I didn't know if any code paths could lead to this. Will remove it.